### DISCUSSION TITLE
Fourier transformation with EmguCV 4

### DISCUSSION THREAD

**Discussion by ELMI5** [emgucv/emgucv#851] [Q&A]
Dear people in the forum,

I would like to do a fourier transform under C# with the latest EmguCV version (4.7.0.5276) and apply filter in the power spectrum.

Is there sample source code for the NEW EmguCV version 4.7 that shows how to Fourier transform and possibly also shows the the power spectrum?

All source codes I found on the net show examples of older EmguCV versions that don't work anymore (probably functions have disappeared or been renamed  during the version upgrade). SIGH!

This code is the best I found, but I never succeeded in "adapting" it to the newest version of EmguCV (4.7.0.5276):

IntPtr complexImage = CvInvoke.cvCreateImage(image.Size,  IplDepth.IplDepth32F, 2);

        CvInvoke.cvSetZero(complexImage);  // Initialize all elements to Zero
        CvInvoke.cvSetImageCOI(complexImage, 1);
        CvInvoke.cvCopy(image, complexImage, IntPtr.Zero);
        CvInvoke.cvSetImageCOI(complexImage, 0);

        Matrix<float> dft = new Matrix<float>(image.Rows, image.Cols, 2);
        CvInvoke.cvDFT(complexImage, dft, Emgu.CV.CvEnum.CV_DXT.CV_DXT_FORWARD, 0);

        //The Real part of the Fourier Transform
        Matrix<float> outReal = new Matrix<float>(image.Size);
        //The imaginary part of the Fourier Transform
        Matrix<float> outIm = new Matrix<float>(image.Size);
        CvInvoke.cvSplit(dft, outReal, outIm, IntPtr.Zero, IntPtr.Zero);

        CvInvoke.cvPow(outReal, outReal, 2.0);
        CvInvoke.cvPow(outIm, outIm, 2.0);

        CvInvoke.cvAdd(outReal, outIm, outReal, IntPtr.Zero);
        CvInvoke.cvPow(outReal, outReal, 0.5);

        CvInvoke.cvAddS(outReal, new MCvScalar(1.0), outReal, IntPtr.Zero); // 1 + Mag
        CvInvoke.cvLog(outReal, outReal); // log(1 + Mag)

        // Swap quadrants
        int cx = outReal.Cols / 2;
        int cy = outReal.Rows / 2;

        Matrix<float> q0 = outReal.GetSubRect(new Rectangle(0, 0, cx, cy));
        Matrix<float> q1 = outReal.GetSubRect(new Rectangle(cx, 0, cx, cy));
        Matrix<float> q2 = outReal.GetSubRect(new Rectangle(0, cy, cx, cy));
        Matrix<float> q3 = outReal.GetSubRect(new Rectangle(cx, cy, cx, cy));
        Matrix<float> tmp = new Matrix<float>(q0.Size);

        q0.CopyTo(tmp);
        q3.CopyTo(q0);
        tmp.CopyTo(q3);
        q1.CopyTo(tmp);                   
        q2.CopyTo(q1);
        tmp.CopyTo(q2);

        CvInvoke.cvNormalize(outReal, outReal, 0.0, 255.0, Emgu.CV.CvEnum.NORM_TYPE.CV_MINMAX, IntPtr.Zero);

        Image<Gray, float> fftImage = new Image<Gray, float>(outReal.Size);
        CvInvoke.cvCopy(outReal, fftImage, IntPtr.Zero);

**Comment by emgucv** [comment]
You can take a look at some of our unit tests that use the CvInvoke.Dft function:

https://github.com/emgucv/emgucv/blob/f671e7dc6cd7a3a02b24e9b94548682d8cedc1df/Emgu.CV.Test/AutoTestImage.cs#L1002

https://github.com/emgucv/emgucv/blob/f671e7dc6cd7a3a02b24e9b94548682d8cedc1df/Emgu.CV.Test/AutoTestImage.cs#L1038

https://github.com/emgucv/emgucv/blob/f671e7dc6cd7a3a02b24e9b94548682d8cedc1df/Emgu.CV.Test/AutoTestImage.cs#L1080

Hope that helps.

**Reply by ELMI5** [reply]
hank you so very much for your hint.
Unfortunately the reverse does not work. It just returns a black image.

I copied your source code and added only the dft(inverse) to check if the original image is returned.
Unfortunately, only a black image is returned. I really just need this start help. This is promised

private void OnFFT(object sender,EventArgs e)
   {
    Mat image = LoadMat("D:\\LENA.jpg", ImreadModes.Grayscale);
    Mat mat = new Mat();
    image.ConvertTo(mat, DepthType.Cv32F);

    Mat complexImage = new Mat(mat.Size, DepthType.Cv32F, 2);
    complexImage.SetTo(new MCvScalar(0, 0));
    CvInvoke.InsertChannel(mat, complexImage, 0);

    Mat dft = new Mat();
   
    CvInvoke.Dft(complexImage, dft, Emgu.CV.CvEnum.DxtType.Forward, 0);
   
   // This is what I added:
    CvInvoke.Dft(dft,complexImage, Emgu.CV.CvEnum.DxtType.InvScale, 0);


    Image<Gray,Byte> imgOut= new Image<Gray, byte>(complexImage.Size);
    complexImage.CopyTo(imgOut);
    picSample.BackgroundImage= imgOut.ToBitmap( );
   }

Any idea?
